### PR TITLE
[InstSimplify] Add basic constant folding for `llvm.sincos`

### PR DESCRIPTION
**Pull request by MacDue**
This calls into the existing constant folding for `llvm.sin` and `llvm.cos`, which currently does not fold for any non-finite values, so most tests are negative tests at the moment.

Note: The constant folding does not consider the `afn` fast-math flag and will produce the same result regardless of if the flag is set.

### CODE REVIEW

**Comment by tschuett** [comment]
Could you add something about `afn`?  IR and/or summery.

**Comment by MacDue** [comment]
> Could you add something about `afn`? IR and/or summery.

Sure, I don't believe constant folding requires `afn` though, right? :slightly_smiling_face:

**Comment by MacDue** [comment]
> I thought about exploiting the flag.

You mean constant folding differently based on if `afn` is set or not?

**Comment by tschuett** [comment]
> > I thought about exploiting the flag.
> 
> You mean constant folding differently based on if `afn` is set or not?

Sure why not. Or alternatively at least stating, we ignore `afn` because ....

File: llvm/test/Transforms/InstSimplify/sincos.ll [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+define { float, float } @sincos_qnam() {
+; CHECK-LABEL: define { float, float } @sincos_qnam() {
+; CHECK-NEXT:    [[RET:%.*]] = call { float, float } @llvm.sincos.f32(float 0x7FF8000000000000)
+; CHECK-NEXT:    ret { float, float } [[RET]]
+;
+  %ret = call { float, float } @llvm.sincos.f32(float 0x7FF8000000000000)
+  ret { float, float } %ret
+}
+
+define { float, float } @sincos_snam() {
```
Comment (arsenm) [inline llvm/test/Transforms/InstSimplify/sincos.ll:118]: typo snam / qnam

File: llvm/test/Transforms/InstSimplify/sincos.ll [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+}
+
+define { float, float } @sincos_neg_zero() {
+; CHECK-LABEL: define { float, float } @sincos_neg_zero() {
+; CHECK-NEXT:    ret { float, float } { float -0.000000e+00, float 1.000000e+00 }
+;
+  %ret = call { float, float } @llvm.sincos.f32(float -0.0)
+  ret { float, float } %ret
+}
+
```
Comment (arsenm) [inline llvm/test/Transforms/InstSimplify/sincos.ll:19]: No tests for normal values?

**Comment by MacDue** [comment]
Kind ping :slightly_smiling_face:

File: llvm/lib/Analysis/ConstantFolding.cpp [hunk 3]
```diff
@@ -3466,6 +3467,44 @@ ConstantFoldStructCall(StringRef Name, Intrinsic::ID IntrinsicID,
...
+      Constant *CosResult =
+          ConstantFoldScalarCall(Name, Intrinsic::cos, TyScalar, Op, TLI, Call);
+      if (!CosResult)
+        return {};
+      return std::make_pair(SinResult, CosResult);
+    };
+
+    if (auto *FVTy = dyn_cast<FixedVectorType>(Ty)) {
+      SmallVector<Constant *, 4> SinResults(FVTy->getNumElements());
+      SmallVector<Constant *, 4> CosResults(FVTy->getNumElements());
```
Comment (arsenm) [inline llvm/lib/Analysis/ConstantFolding.cpp:3489]: Use the default small size? Or 8 is more common for vectors

File: llvm/test/Transforms/InstSimplify/sincos.ll [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+define { <vscale x 2 x float>, <vscale x 2 x float> } @sincos_zero_scalable_vector() {
+; CHECK-LABEL: define { <vscale x 2 x float>, <vscale x 2 x float> } @sincos_zero_scalable_vector() {
+; CHECK-NEXT:    [[RET:%.*]] = call { <vscale x 2 x float>, <vscale x 2 x float> } @llvm.sincos.nxv2f32(<vscale x 2 x float> zeroinitializer)
+; CHECK-NEXT:    ret { <vscale x 2 x float>, <vscale x 2 x float> } [[RET]]
+;
+  %ret = call { <vscale x 2 x float>, <vscale x 2 x float> } @llvm.sincos.nxv2f32(<vscale x 2 x float> zeroinitializer)
+  ret { <vscale x 2 x float>, <vscale x 2 x float> } %ret
+}
```
Comment (arsenm) [inline llvm/test/Transforms/InstSimplify/sincos.ll:112]: In a follow up, should probably make sure the zero, undef, and splat cases work for scalable vectors

File: llvm/lib/Analysis/ConstantFolding.cpp [hunk 5]
```diff
@@ -3466,6 +3467,44 @@ ConstantFoldStructCall(StringRef Name, Intrinsic::ID IntrinsicID,
...
+
+      for (unsigned I = 0, E = FVTy->getNumElements(); I != E; ++I) {
+        Constant *Lane = Operands[0]->getAggregateElement(I);
+        std::tie(SinResults[I], CosResults[I]) =
+            ConstantFoldScalarSincosCall(Lane);
+        if (!SinResults[I])
+          return nullptr;
```
Comment (arsenm) [inline llvm/lib/Analysis/ConstantFolding.cpp:3496]: Inside ConstantFoldScalarSincosCall, you assume they can fail independently. Are there really any cases where one can fail and not the other? In any case, the error handling should be consistent in both places

Comment (MacDue) [reply llvm/lib/Analysis/ConstantFolding.cpp]: > Are there really any cases where one can fail and not the other?

Probably not, but they both look like calls that could return null, so checking them both seems justified.

**Review by arsenm — APPROVED** [review]

**Comment by MacDue** [comment]
Reverted this for now (https://github.com/llvm/llvm-project/pull/119149). I'll look into the buildbot issues and reland tomorrow.

**Pull request merged**